### ISSUE TITLE
Support for clipboard=autoselect behaviour

### ISSUE THREAD

**Issue by yshui** [neovim/neovim#2325]
It will be convenient if neovim can automatically send visual selection to clipboard (without yanking).

**Comment by tarruda** [comment]
@bfredl will this be possible after #2077?

**Comment by bfredl** [comment]
Not in #2077 #2244, but I plan to work on it after those are done (they are quite big PR:s already)

**Comment by pickfire** [comment]
+1, I hope I can use `set mouse=a` to copy and paste to another application.

**Comment by fmoralesc** [comment]
I think we should prioritize this for the next release, I've seen many reports about this: https://www.youtube.com/watch?v=yAqj8jupFws and [here](https://www.reddit.com/r/neovim/comments/3s8n32/how_to_copy_on_selection/). I think people complain about us defaulting to `mouse=a` because of this aswell.

**Comment by pickfire** [comment]
@fmoralesc I think setting `mouse` to anything instead of `mouse=a` is better than setting `mouse=a` because most people need to copy and paste it. It will also make mouse highlight ugly:

![2015-11-12-101818_1600x900_scrot](https://cloud.githubusercontent.com/assets/4687791/11108993/f82d458a-8926-11e5-9643-6f0dc65a1bb9.png)

**Comment by bfredl** [comment]
I hope to look into this the following days. I have read the vim implemention, but it relies heavily on the fact that vim "owns" the X11 selection, so that "updating" the selection is virtually effortless. We'll need to either invoke `xsel` on every visual mode keypress, or somehow detect reliably when nvim loses focus and update the selection then.

**Comment by bfredl** [comment]
Wow. When in commandline mode, vim actually emulates the "dumb" terminal selection (I mean selecting line numbers and terminal whitespace and everything), I always thought it just turned off mouse handling and let the terminal do it.

**Comment by fmoralesc** [comment]
@bfredl Wow indeed. Do you plan to remove that?

**Comment by bfredl** [comment]
@fmoralesc How, from vim? :) Well, I will start with visual mode selection as I suppose that whats most useful and simplest to implement. 
Implementing `autoselectml` will probably require a lot more code written or ported from vim. Actually in cmdline mode in vim you can autoselect anything _but_ the command line, and in command window mode `autoselect` and `autoselectml` seems to interfere with each other.

**Comment by fmoralesc** [comment]
@bfredl LOL, I guess I wasn't paying enough attention :p

**Comment by bfredl** [comment]
Hmm, what actually happens in vim, when visual mode and another x11 app wants to paste selection, is to save editor state,literally do `"*y` and then restore editor state  so that it seems nothing actually happened (expect for updating * yankreg). Not sure if it simplest for us to do that too or to extract function to determine visual text contents without change any state...

**Comment by bfredl** [comment]
It seems like nvim doesn't distinguish between mouse drag and mouse release events, at least not from observing `do_mouse` (I get first `is_click, is_drag = true,false` then `false,true` but never `false,false`). Just by testing with `ttest` the terminal do emit release events, but they aren't passed on in nvim somehow.

**Comment by bfredl** [comment]
Seems regression from vim, just try `:map <LeftRelease> anycommand` works in `vim` but not `nvim`. Or select window by clicking on statusline (and not dragging).

**Comment by lqueryvg** [comment]
+1 for this feature (clipboard=autoselect behaviour)

It's the main feature preventing me moving from vim to neovim.

I work on a linux desktop and expect to be able to select & paste using mouse left & middle buttons only (no keyboard). This works fine with vim (with guioptions=a & mouse=a), but sadly not with neovim.

@bfredl you said:

> invoke xsel on every visual mode keypress

I don't think visual keyboard selection is such an issue. If I make a visual selection with the keyboard, I expect to have to yank (with the keyboard) to get it into the clipboard.

As you say:
> the case of mouse selection is simpler, just update when the mouse is released

I'm hoping this is still possible ?

**Comment by fmoralesc** [comment]
@lqueryvg, bfredl is now working on reimplementing the clipboard code without using the providers interface, see #4523.

**Comment by bfredl** [comment]
Last time I checked, `<LeftRelease>` event (which would have allowed a swift resolution of the main issue) was broken :(. Time to check again...

**Comment by bfredl** [comment]
(I mean one could have done `vmap <LeftRelease> "*ygv` (or a builtin equivalent thereof) and basically be done with it...)

**Comment by lqueryvg** [comment]
If you need any help testing this, let me know.

**Comment by xeliba** [comment]
Until this functionality is really built in via 'autoselect' one can still press the SHIFT key while using the mouse for selections. While SHIFT is pressed, vim/nvim does not capture mouse button events but lets the terminal handle them. So there is no need to turn off 'set mouse=a'.

See `:help mouse`  in the vim/nvim docu.

**Comment by lqueryvg** [comment]
@xeliba, it's true that most terminal emulators handle the mouse events themselves when you hold down `Shift`, but that's not much help when working with vim splits or line numbering.

**Comment by kierun** [comment]
If you need any help testing this, let me know.

This is the feature which is blocking me to using neovim!

**Comment by bfredl** [comment]
Good to know :+1: 

#3708 might work in the terminal in the meanwhile (even if very inefficient, and needs a rebase), but I hope to get #4523 in a testable/usable (but probably not mergable) state in the weekend or so.

**Comment by kierun** [comment]
Let me clarify.  I came to this bug via the [neovim-qt one](https://github.com/equalsraf/neovim-qt/issues/111). In my not so odd (hopefully⸮) set up, copy-and-paste works fine from the console. However, neovim-qt has no copy-and-paste at all.  The above fix did work for neovim-qt and has not broken/changed the console behaviour.

**Comment by waiting-for-dev** [comment]
+1 . I found this issue because it breaks visual mappings that rely on `@*`. `SHIFT` workaround is not useful in this scenario (but it is indeed in others!). This is the only handicap I have found while migration to neovim, so I have to say: great job!!! Please, if you need any help testing in different environments tell me (in my case, Linux Fedora).

**Comment by antoinemadec** [comment]
@bfredl is there any progress on this feature ?
I would like to copy and paste from/to the clipboard with `set mouse=a`, I tried all combinations of `set clipboard` but it does not work.
I also tried to use this workaround: `vmap <LeftRelease> "*ygv`, but it does not work with `set mouse=n` or `set mouse=a`.

For info, I am compiling nvim with the latest sources.
Thanks in advance, and thanks for working on this amazing project :)

**Comment by lilydjwg** [comment]
I tried to switch to neovim again, and quickly encountered this. I'm still waiting for this to switch.

**Comment by kierun** [comment]
@lilydjwg The solution [here](https://github.com/equalsraf/neovim-qt/issues/111) does work fine. It is a work around but a simple one.

**Comment by lilydjwg** [comment]
@kierun What about selecting with keyboard? Like ve vaw vi" vap V etc.

**Comment by kierun** [comment]
@lilydjwg It works fine for me. ☺ Try it out, it should take you less than five minutes to see if it is suitable or annoying for you.

**Comment by lilydjwg** [comment]
@kierun No it doesn't work when I use keyboard to select.

**Comment by lqueryvg** [comment]
With all this talk of keyboard mappings, I hope the focus isn't being changed on this issue?
Any kind of keyboard mappings are *not* solutions or even workarounds for this.

With the setting described by the OP, it should be possible to select and copy with the mouse only - without touching the keyboard.

I thought Neovim was supposed to be a refactoring (plus extensions and improvements). I hope it doesn't intend to remove fundamental features like this one.

**Comment by lilydjwg** [comment]
No, the focus is X primary selection unchanged. Not only selections created with the mouse; all selections should be available over the X primary selection mechanism.

I didn't consider setting `*` as the default register, but even that way I still need to press an extra `y` each time. It is different than all other X programs (and neovim itself with mouse selections).

**Comment by LunarLambda** [comment]
Any updates on this?

**Comment by justinmk** [comment]
As mentioned above, here is a workaround:

```
set mouse=a
vmap <LeftRelease> "*ygv
```

I tested this on NVIM v0.3.2 87d67814e with Terminal.app. It (obviously) requires a working clipboard provider. Run `:checkhealth` to see if your clipboard provider is working.

**Comment by zeertzjq** [comment]
More workarounds have been mentioned in #27675.
